File: leetcode/54.spiral-matrix.py
# @leet start
class Solution:
    def spiralOrder(self, matrix: list[list[int]]) -> list[int]:
        res: list[int] = []

        left = 0
        right = len(matrix[0]) - 1
        up = 0
        down = len(matrix) - 1

        while left <= right and up <= down:
            for i in range(left, right + 1):
                res.append(matrix[up][i])

            for i in range(up + 1, down + 1):
                res.append(matrix[i][right])

            if left < right and up < down:
                for i in range(right - 1, left - 1, -1):
                    res.append(matrix[down][i])

                for i in range(down - 1, up, -1):
                    res.append(matrix[i][left])

            left += 1
            right -= 1
            up += 1
            down -= 1

        return res


# @leet end

File: leetcode/54.spiral-matrix.ts
// @leet start
function spiralOrder(matrix: number[][]): number[] {
  const result: number[] = [];

  let startCol = 0;
  let endCol = matrix[0].length - 1;
  let startRow = 0;
  let endRow = matrix.length - 1;

  while (startCol <= endCol && startRow <= endRow) {
    for (let col = startCol; col <= endCol; col++) {
      result.push(matrix[startRow][col]);
    }

    for (let row = startRow + 1; row <= endRow; row++) {
      result.push(matrix[row][endCol]);
    }

    // If we're only 1D don't backtrack
    if (startCol === endCol || startRow === endRow) {
      break;
    }

    for (let col = endCol - 1; col >= startCol; col--) {
      result.push(matrix[endRow][col]);
    }

    for (let row = endRow - 1; row > startRow; row--) {
      result.push(matrix[row][startCol]);
    }

    startCol++;
    endCol--;
    startRow++;
    endRow--;
  }

  return result;
}
// @leet end
